### PR TITLE
Powderhouse subsystems and directives

### PR DESCRIPTION
**Pull request by KathleenDollard**
This addresses some of the issues raised in #2342 regarding directives (not response files environment variables(..

Remains in draft because: 

* There are more changes to the Tokenizer than I expected to evaluate directives separately - these changes extended to several files
* I put some data onto the CLI Configuration, because I didn't know where to put it. When we determine how to carry information, we need to plan a way to tuck it away, probably similar to the SubSystem static class for subsystems. 
* A major piece of information is where the Tokenizer should begin. There are two reasons for this - the exe name may or may not be included in the args array and we need to deal with this, and Initialize methods for subsystems may have handled some of the args. We are limiting that handling to the start of the args array.
* There are numerous naming and other issues to resolve, but I wanted to get the basics out for folks to look at before I am mostly busy next week.

### CODE REVIEW

**Comment by KathleenDollard** [comment]
This issue has files changed that I did not expect. I need to review this with Jean or Mikayla, regardless of content, and I know it makes it hard to review. So, unless you are in a patient mood, you can wait for a cleaner PR. 

I did not deliberately delete most of the code marked as deleted.

**Review by JeanJoeris — COMMENTED** [review]
Paired with @KathleenDollard reviewing this. Left comments from review. Overall looks good, big remaining questions this review does not address are largely around:
- `StringExtensions`
  - haven't reviewed as it is outdated / in flux
- Handling of Directives as options
  - conversations around this are currently in flight

File: src/System.CommandLine.Subsystems.Tests/TestData.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        private readonly List<object[]> _data =
+        [
+            ["--version", true],
+            ["-v", true],
+            ["-vx", true],
+            ["-xv", true],
+            ["-x", false],
+            [null, false],
+            ["", false],
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems.Tests/TestData.cs:26]: Comment added several places, not in existing functional tests

File: src/System.CommandLine.Subsystems.Tests/TestData.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            ["[diagram] -v", true],
+            ["[diagram] x -v", true ],
+            ["[diagramX]", false],
+            ["[diagram] [other]", true],
+            ["[diagram:hello] [other] [and-also]", true],
+            ["x", false],
+            ["-o", false],
+            ["x -x", false],
+            [null, false],
+            ["", false]
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems.Tests/TestData.cs]: Comment added

File: src/System.CommandLine.Subsystems/Directives/DiagramSubsystem.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            /*
+            case DirectiveResult { Directive: not DiagramDirective }:
+                break;
+            */
+
+            // TODO: This logic is deeply tied to internal types/properties. These aren't things we probably want to expose like SymbolNode. See #2349 for alternatives
+            /*
+            case ArgumentResult argumentResult:
+                {
+                    var includeArgumentName =
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems/Directives/DiagramSubsystem.cs:77]: I think this problem is gnarly enough to include it here

File: src/System.CommandLine.Subsystems/Directives/DirectiveOption.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.CommandLine.Directives
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems/Directives/DirectiveOption.cs]: Done

File: src/System.CommandLine.Subsystems/Directives/DirectiveResult.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.CommandLine.Directives
+{
+    public class DirectiveResult
+    {
```
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems/Directives/DirectiveResult.cs]: Done

File: src/System.CommandLine.Subsystems/Directives/DiagramSubsystem.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (KathleenDollard) [reply src/System.CommandLine.Subsystems/Directives/DiagramSubsystem.cs:77]: @mhutch may have an opinion here

**Comment by KathleenDollard** [comment]
@KalleOlaviNiemitalo 

> The changes might be easier to understand if you updated the API compatibility test files too.

TBH, we are still getting everything to work with the split approach and have not given much thought to the API yet. We will do that a bit later. 

I think the new subsystem approach will take some docs beyond API, and we're working on some design docs.

File: src/System.CommandLine.Subsystems/Subsystems/InitializationContext.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.CommandLine.Subsystems;
+
+public class InitializationContext(CliConfiguration configuration, IReadOnlyList<string> args)
```
Comment (patriksvensson) [inline src/System.CommandLine.Subsystems/Subsystems/InitializationContext.cs:6]: Isn't the name of this class a bit too general to be public?

**Review by baronfel — APPROVED** [review]

**Pull request merged**